src/core/config/dns_client.h

```diff
@@ -168,7 +168,7 @@
  *
  */
 #ifndef OPENTHREAD_CONFIG_DNS_CLIENT_DEFAULT_SERVICE_MODE
-#define OPENTHREAD_CONFIG_DNS_CLIENT_DEFAULT_SERVICE_MODE OT_DNS_SERVICE_MODE_SRV_TXT_OPTIMIZE
+#define OPENTHREAD_CONFIG_DNS_CLIENT_DEFAULT_SERVICE_MODE OT_DNS_SERVICE_MODE_SRV_TXT_SEPARATE
```

Comment:
Changing this default will affect the default behavior for all OpenThread devices. Are we sure we want to do this?

If the goal is simply to support testing "separate" mode, but that devices in production should default to "optimize" mode, should we leave this unchanged and only make the change for given test modes?

@abtink @EskoDijk thoughts?

Reply:
Using the default "separate" could have the advantage to be compliant to (most likely) new rules defined in https://datatracker.ietf.org/doc/html/draft-ietf-dnsop-qdcount-is-one-03 , which disallows a query with multiple questions.

In practice OpenThread-based BRs will accept a multi-question query, making the entire process more efficient. There are also BRs that don't accept a multi-question query. In this case, if the `OPTIMIZE` mode would then switches to `SEPARATE` for future queries that would be helpful.  I don't think the current OT code does this yet? It will, for each new DNS query, probably try to send an optimized 2-question query and if that fails convert it to two single-question queries (SRV/TXT) as far as I can see in the code.  This becomes then less efficient than the `SEPARATE` mode if such a BR is in the home network.  Also for external DNS queries (outside the mesh), the `SEPARATE` mode is for sure best.

That's my thoughts, but it's not an answer yet if we should make this change or not :) 
It seems good to avoid trouble later in the future and keep things simple, at the expense of some efficiency in networks that use OpenThread-BRs anyway.

Reply:
On one hand, I do like the "optimize" mode, which first attempts to query for both SRV and TXT together, and only if the server doesn't respond (or responds with an error) does it send the queries separately. This reduces traffic, particularly considering that a SED device may be the originator of the DNS queries.

On the other hand, I'm aware of the discussions regarding enforcing a query count of one. I've always found it inefficient to burden a less capable device to accommodate BRs and servers, which are typically computationally more capable.

> In this case, if the OPTIMIZE mode would then switches to SEPARATE for future queries that would be helpful. I don't think the current OT code does this yet?

I was thinking along the same lines. We cam later add an enhancement where we use "separate" service mode as long as the server address remains the same and we've learned from a previous interaction that the server cannot handle two questions.

----

Here is my proposal:

- We keep the existing default configuration as "optimize" mode (users can, of course, change this at runtime or even per query if they wish).
- I can later implement the enhancement to remember whether the currently selected server can handle two questions and override "optimize" requests accordingly.

Reply:
Submitted
- https://github.com/openthread/openthread/pull/10403

Reply:
Just an additional thought here: by removing support for the `OPTIMIZE` mode we would reduce code complexity and save several 100s of bytes in code size.  If it's not certain that all BRs would support it, and if external DNS servers (typically?) don't support it, and future specification will require a server not to support it, is it still worth doing?

Reply:
I'm not sure if there'd be any significant code size reduction  by removing "optimize" mode support, as the implementation is already quite lean.

The main complexity in the code stems from needing multiple queries for the same request. We have to prepare and send multiple queries, wait for potentially out-of-order responses, and only then report the collected information to the user. PR #8772, which added multi-question support, increased the code size by approximately 1100 bytes. 

> if it's not certain that all BRs would support it, and if external DNS servers (typically?) don't support it

I expect the typical use case for "service resolution" to be within the local domain, where devices send the query to a Border Router as the resolver. This resolver would then act as a discovery proxy, performing mDNS query on the AIL, which allows for multiple questions.

Reply:
In my "100s of bytes" I do also include the #10403 PR, which has a size in the 100s (e.g. +368), because my feeling was that this PR would be required in order to make the multi-question approach efficient. For example, some home networks will have only Thread BRs that don't support multi-question. Which means (without that PR) that each initial service query would fail, I think? (I may be wrong here.)

Relevant to note for this discussion is the IETF specs coming up on how to ask multiple questions using an option, see https://datatracker.ietf.org/doc/html/draft-ietf-dnssd-multi-qtypes .  This may still take a while but at some time in the future this would be the preferred method. But not something we can rely on today.

So both have pros and cons so also okay for me to leave the current default service mode to "optimize".

Reply:
Closing this PR for now (keeping the default settings as "optimized"). Thanks for all the discussions. We can consider revisiting this to incorporate any new insights we've gained.